File: Sources/BraintreeApplePay/BTApplePayCardNonce.m
#if __has_include(<Braintree/BraintreeApplePay.h>)
#import <Braintree/BTApplePayCardNonce.h>
#else
#import <BraintreeApplePay/BTApplePayCardNonce.h>
#endif

@implementation BTApplePayCardNonce

- (instancetype)initWithJSON:(BTJSON *)json {
    NSString *cardType = [json[@"details"][@"cardType"] asString] ?: @"ApplePayCard";
    self = [super initWithNonce:[json[@"nonce"] asString] type:cardType isDefault:[json[@"default"] isTrue]];
    
    if (self) {
        _isDeviceToken = [json[@"details"][@"isDeviceToken"] isTrue];
        _binData = [[BTBinData alloc] initWithJSON:json[@"binData"]];
    }
    return self;
}

@end

File: Sources/BraintreeApplePay/BTApplePayCardNonce.swift
import Foundation

#if canImport(BraintreeCore)
import BraintreeCore
#endif

/// Contains information about a tokenized Apple Pay card.
@objcMembers public class BTApplePayCardNonce: BTPaymentMethodNonce {

    /// The BIN data for the card number associated with this nonce.
    public let binData: BTBinData

    /// This Boolean indicates whether this tokenized card is a device-specific account number (DPAN) or merchant/cloud token (MPAN). Available on iOS 16+.
    /// If `isDeviceToken` is `false`, then token type is MPAN
    public var isDeviceToken: Bool

    init?(json: BTJSON) {
        guard let nonce = json["nonce"].asString() else { return nil }

        let cardType = json["details"]["cardType"].asString() ?? "ApplePayCard"
        let isDefault = json["default"].isTrue

        self.isDeviceToken = json["details"]["isDeviceToken"].asBool() ?? true

        binData = BTBinData(json: json["binData"])
        super.init(nonce: nonce, type: cardType, isDefault: isDefault)
    }
}
